xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-api/src/main/java/org/xwiki/livedata/internal/LiveDataRenderer.java

```diff
@@ -0,0 +1,118 @@
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.livedata.internal;
+
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.StringUtils;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.livedata.LiveDataConfiguration;
+import org.xwiki.livedata.LiveDataConfigurationResolver;
+import org.xwiki.livedata.LiveDataException;
+import org.xwiki.rendering.block.Block;
+import org.xwiki.rendering.block.GroupBlock;
+import org.xwiki.security.authorization.ContextualAuthorizationManager;
+import org.xwiki.skinx.SkinExtension;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import static java.util.Collections.singletonList;
+import static java.util.Collections.singletonMap;
+import static org.xwiki.security.authorization.Right.SCRIPT;
+
+/**
+ * Renders a Live Data. Can be called from different context, in particular from the Live Data macro and the Live Data
+ * script service.
+ *
+ * @version $Id$
+ * @since 16.0.0RC1
+ */
+@Component(roles = LiveDataRenderer.class)
+@Singleton
+public class LiveDataRenderer
+{
+    /**
+     * Used to add default Live Data configuration values.
+     */
+    @Inject
+    private LiveDataConfigurationResolver<LiveDataConfiguration> defaultLiveDataConfigResolver;
+
+    @Inject
+    private LiveDataRendererConfiguration liveDataRendererConfiguration;
+
+    @Inject
+    private ContextualAuthorizationManager contextualAuthorizationManager;
+
+    /**
+     * The component used to load the JavaScript code of the Live Data widget.
+     */
+    @Inject
+    @Named("jsfx")
+    private SkinExtension jsfx;
+
+    /**
+     * Render the Live Data content.
+     *
+     * @param parameters the parameters used to render the Live Data
+     * @param advancedParameters a json string containing the advanced parameters of the Live Data
+     * @param restricted when {@code true}, the Live Data is contained in a restricted environment and its content
+     *     will be considered as untrusted
+     * @return the {@link Block}s to render for this Live Data
+     * @throws LiveDataException in case of issue when resolving the Live Data configuration
+     */
+    public List<Block> render(LiveDataRendererParameters parameters, String advancedParameters, boolean restricted)
```

Comment:
Why is this returning a list of blocks? From what I can see this only complicates things as the return value is always a list with a single item, and it also has the consequence that any caller now needs to iterate over the blocks. If there shall be several blocks in the feature, a composite block could always be used to group them transparently.

Reply:
Good catch. This will allow to simplify the macro code as well. Thanks.